File: src/engine/adapters/sqlite/definition.ts
import type { ConstraintMeta, ObjectDefinition } from '../../../shared/domain/definition';
import { decodePath, encodePath, type NodePath } from '../../../shared/domain/tree';
import { AdapterError } from '../errors';
import type { QueryExecutor, ReadTarget } from './catalog';
import * as catalog from './catalog';
import { quoteIdent } from './read';

export type RelationLikeKind = 'table' | 'view';

interface SqliteMasterRow {
  sql: string | null;
}

// SHOW CREATE TABLE has no SQLite analogue — sqlite_master.sql already *is* the CREATE statement
// as the user (or a prior migration) wrote it, verbatim, so this is "asked, never composed" in
// its simplest possible form (the position mysql-family/definition.ts:97-101 takes about
// SHOW CREATE TABLE — SQLite just skips the round trip through the server entirely).
function stripOneTrailingSemicolon(text: string): string {
  const match = /;\s*$/.exec(text);
  return match ? text.slice(0, text.length - match[0].length) : text;
}

// F19/D24: SQLite has no CHECK-constraint catalog at all — a CHECK is visible only inside the
// CREATE statement's own text, which the Source pane already shows verbatim. Listing PK/UNIQUE/FK
// here (composed from the same pragmas describe() uses) and saying so for CHECK beats an empty
// Constraints section that looks like there simply are none.
function buildConstraints(
  exec: QueryExecutor,
  schema: string,
  table: string,
  target: ReadTarget,
): ConstraintMeta[] {
  const constraints: ConstraintMeta[] = [];
  if (target.primaryKey && target.primaryKey.length > 0) {
    constraints.push({
      name: `${table}_pkey`,
      type: 'primaryKey',
      definition: `(${target.primaryKey.map(quoteIdent).join(', ')})`,
    });
  }

  for (const idx of catalog.listIndexes(exec, table)) {
    if (idx.primary || !idx.unique) continue; // PK covered above; a plain index isn't a constraint
    constraints.push({
      name: idx.name,
      type: 'unique',
      definition: `(${idx.columns.map(quoteIdent).join(', ')})`,
    });
  }

  for (const fk of catalog.listForeignKeys(exec, schema, table)) {
    const refTable = decodePath('', fk.referencedPath).segments.at(-1)?.name ?? '';
    constraints.push({
      name: fk.name,
      type: 'foreignKey',
      definition:
        `(${fk.columns.map(quoteIdent).join(', ')}) REFERENCES ${quoteIdent(refTable)} ` +
        `(${fk.referencedColumns.map(quoteIdent).join(', ')})`,
    });
  }

  return constraints;
}

export function buildDefinition(
  exec: QueryExecutor,
  segments: NodePath['segments'],
  schema: string,
  object: { kind: RelationLikeKind; name: string },
): ObjectDefinition {
  const masterType = object.kind === 'view' ? 'view' : 'table';
  const rows = exec<SqliteMasterRow>('SELECT sql FROM sqlite_master WHERE type = ? AND name = ?', [
    masterType,
    object.name,
  ]);
  const raw = rows[0]?.sql;
  if (typeof raw !== 'string' || raw.length === 0) {
    throw new AdapterError(
      'E_QUERY',
      `sqlite_master returned no definition for "${schema}"."${object.name}"`,
    );
  }
  const statement = stripOneTrailingSemicolon(raw);

  let constraints: ConstraintMeta[] = [];
  let notes: string[] = [];
  if (object.kind === 'table') {
    const target = catalog.getReadTarget(exec, schema, object.name);
    constraints = buildConstraints(exec, schema, object.name, target);
    notes = [
      'CHECK constraints, if any, appear only in the Source text above — SQLite has no separate ' +
        'catalog for them.',
    ];
  }

  return {
    path: encodePath(segments),
    kind: object.kind,
    qualifiedName: `${schema}.${object.name}`,
    statements: [statement],
    language: 'sql',
    origin: 'server',
    notes,
    constraints,
    documentSchema: null,
    sections: [],
    generatedAt: new Date().toISOString(),
  };
}

File: shell/internal/adapters/sqlite/definition.go
package sqlite

import (
	"database/sql"
	"strings"
	"time"

	"github.com/kirathecat/kira-studio/shell/internal/adapters"
	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

// buildConstraints is definition.ts's own — F19/D24: SQLite has no CHECK-constraint catalog at
// all; a CHECK is visible only inside the CREATE statement's own text, which the Source pane
// already shows verbatim. Listing PK/UNIQUE/FK here (composed from the same pragmas describe()
// uses) and saying so for CHECK beats an empty Constraints section that looks like there simply
// are none.
func buildConstraints(exec QueryExecutor, schema, table string, target ReadTarget) ([]model.ConstraintMeta, error) {
	var constraints []model.ConstraintMeta
	if len(target.PrimaryKey) > 0 {
		quoted := make([]string, len(target.PrimaryKey))
		for i, c := range target.PrimaryKey {
			quoted[i] = quoteIdent(c)
		}
		constraints = append(constraints, model.ConstraintMeta{
			Name: table + "_pkey", Type: "primaryKey",
			Definition: "(" + strings.Join(quoted, ", ") + ")",
		})
	}

	indexes, err := listIndexes(exec, table)
	if err != nil {
		return nil, err
	}
	for _, idx := range indexes {
		if idx.Primary || !idx.Unique { // PK covered above; a plain index isn't a constraint
			continue
		}
		quoted := make([]string, len(idx.Columns))
		for i, c := range idx.Columns {
			quoted[i] = quoteIdent(c)
		}
		constraints = append(constraints, model.ConstraintMeta{
			Name: idx.Name, Type: "unique", Definition: "(" + strings.Join(quoted, ", ") + ")",
		})
	}

	fks, err := listForeignKeys(exec, schema, table)
	if err != nil {
		return nil, err
	}
	for _, fk := range fks {
		refPath, err := model.DecodePath("", fk.ReferencedPath)
		refTable := ""
		if err == nil && len(refPath.Segments) > 0 {
			refTable = refPath.Segments[len(refPath.Segments)-1].Name
		}
		quotedCols := make([]string, len(fk.Columns))
		for i, c := range fk.Columns {
			quotedCols[i] = quoteIdent(c)
		}
		quotedRefCols := make([]string, len(fk.ReferencedColumns))
		for i, c := range fk.ReferencedColumns {
			quotedRefCols[i] = quoteIdent(c)
		}
		constraints = append(constraints, model.ConstraintMeta{
			Name: fk.Name, Type: "foreignKey",
			Definition: "(" + strings.Join(quotedCols, ", ") + ") REFERENCES " + quoteIdent(refTable) +
				" (" + strings.Join(quotedRefCols, ", ") + ")",
		})
	}

	return constraints, nil
}

// buildDefinition is definition.ts's own. SHOW CREATE TABLE has no SQLite analogue —
// sqlite_master.sql already *is* the CREATE statement as the user (or a prior migration) wrote it,
// verbatim, so this is "asked, never composed" in its simplest possible form.
func buildDefinition(exec QueryExecutor, segments []model.PathSegment, schema string, kind, name string) (model.ObjectDefinition, error) {
	masterType := "table"
	if kind == "view" {
		masterType = "view"
	}
	var raw sql.NullString
	found := false
	err := exec("SELECT sql FROM sqlite_master WHERE type = ? AND name = ?", []any{masterType, name}, func(r *sql.Rows) error {
		found = true
		return r.Scan(&raw)
	})
	if err != nil {
		return model.ObjectDefinition{}, err
	}
	if !found || !raw.Valid || raw.String == "" {
		return model.ObjectDefinition{}, adapters.New(adapters.CodeQuery, `sqlite_master returned no definition for "`+schema+`"."`+name+`"`, nil)
	}
	statement := adapters.StripOneTrailingSemicolon(raw.String)

	var constraints []model.ConstraintMeta
	var notes []string
	if kind == "table" {
		target, err := getReadTarget(exec, schema, name)
		if err != nil {
			return model.ObjectDefinition{}, err
		}
		constraints, err = buildConstraints(exec, schema, name, target)
		if err != nil {
			return model.ObjectDefinition{}, err
		}
		notes = []string{
			"CHECK constraints, if any, appear only in the Source text above — SQLite has no separate " +
				"catalog for them.",
		}
	}

	return model.ObjectDefinition{
		Path: model.EncodePath(segments), Kind: kind, QualifiedName: schema + "." + name,
		Statements: []string{statement}, Language: "sql", Origin: "server", Notes: notes,
		Constraints: constraints, DocumentSchema: nil, Sections: []model.DefinitionSection{},
		GeneratedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}, nil
}
